src/compiler/checker.ts

```diff
@@ -59,7 +59,20 @@ namespace ts {
             isArgumentsSymbol: symbol => symbol === argumentsSymbol,
             getDiagnostics,
             getGlobalDiagnostics,
-            getTypeOfSymbolAtLocation,
+
+            // Get the narrowed type of symbol at given location instead of just getting
+            // the type of the symbol.
+            // eg.
+            // function foo(a: string | number) {
+            //     if (typeof a === "string") {
+            //         a/**/
+            //     }
+            // }
+            // getTypeOfSymbol for a would return type of parameter symbol string | number
+            // Unless we provide location /**/, checker wouldn't know how to narrow the type
+            // By using getNarrowedTypeOfSymbol would return string since it would be able to narrow
+            // it by typeguard in the if true condition

```

Comment:
I changed it to
```

            // The language service will always care about the narrowed type of a symbol, because that is
            // the type the language says the symbol should have.
```
I think it's pretty obvious that we need the narrowed type here.